api/kv.go

```diff
@@ -0,0 +1,328 @@
+//
+// The mount path is the location where the target KV secrets engine resides
+// in Vault. The version refers to the version of the target KV secrets engine.
+//
+// Vault development servers tend to have "secret" as the mount path
+// and 2 as the version, as these are the default settings when a server
+// is started in -dev mode.
+//
+// Learn more about the KV secrets engine here:
+// https://www.vaultproject.io/docs/secrets/kv
+type KVClient struct {
+	c         *Client
+	mountPath string
+	version   int
+}
+
+type KVSecret struct {
+	Data     map[string]interface{}
+	Metadata *VersionMetadata
+	Raw      *Secret
+}
+
+type KVMetadata struct {
+	CASRequired        bool                   `mapstructure:"cas_required"`
+	CreatedTime        time.Time              `mapstructure:"created_time"`
+	CurrentVersion     int                    `mapstructure:"current_version"`
+	CustomMetadata     map[string]interface{} `mapstructure:"custom_metadata"`
+	DeleteVersionAfter time.Duration          `mapstructure:"delete_version_after"`
+	MaxVersions        int                    `mapstructure:"max_versions"`
+	OldestVersion      int                    `mapstructure:"oldest_version"`
+	UpdatedTime        time.Time              `mapstructure:"updated_time"`
+	// Keys are stringified ints, e.g. "3"
+	Versions map[string]VersionMetadata `mapstructure:"versions"`
+}
+
+type VersionMetadata struct {
+	Version      int       `mapstructure:"version"`
+	CreatedTime  time.Time `mapstructure:"created_time"`
+	DeletionTime time.Time `mapstructure:"deletion_time"`
+	Destroyed    bool      `mapstructure:"destroyed"`
+	// There is currently no version-specific custom metadata.
+	// This field is just a copy of what's in the CustomMetadata field
+	// for the full KVMetadata of the secret.
+	CustomMetadata map[string]string `mapstructure:"custom_metadata"`
+}
+
+func (c *Client) KVv1(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 1}
+}
+
+func (c *Client) KVv2(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 2}
+}
+
+func (kv *KVClient) Read(ctx context.Context, secretPath string) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, fmt.Errorf("error encountered while reading secret at %s: %v", pathToRead, err)
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+// ReadVersion returns the data and metadata for a specific version of the
+// given secret. If that version has been deleted, the Data field on the
+// returned secret will be nil, and the Metadata field will contain the deletion time.
```

Comment:
It may be helpful to point the users to which function they should call to get the list of available versions.